### ISSUE TITLE
Questions about training details of CodeBERT/GraphCodeBERT on clone detection task 

### ISSUE THREAD

**Issue by zzxn** [microsoft/CodeBERT#93]
A few questions about clone detection:

* Q1: How is CodeBERT finetuned? There is no code, and I cannot reproduce the reported results by adapting the code of GraphCodeBERT.
* Q2: The paper says that "The probability of true clone is calculated by dot product from the representation of [CLS]." But in the code here, the probability is actually calculated by concatenating two [CLS] representations then feeding them into MLP. Which one is correct?  Or my version of the paper is outdated? 

Thanks and look forward to your reply.

**Comment by tyhxyz** [comment]
same question,follwing code shows the probability is actually calculated by concatenating two [CLS] representations then feeding them into MLP
` def forward(self, inputs_ids_1,position_idx_1,attn_mask_1,inputs_ids_2,position_idx_2,attn_mask_2,labels=None): 
        bs,l=inputs_ids_1.size()
        inputs_ids=torch.cat((inputs_ids_1.unsqueeze(1),inputs_ids_2.unsqueeze(1)),1).view(bs*2,l)
        position_idx=torch.cat((position_idx_1.unsqueeze(1),position_idx_2.unsqueeze(1)),1).view(bs*2,l)
        attn_mask=torch.cat((attn_mask_1.unsqueeze(1),attn_mask_2.unsqueeze(1)),1).view(bs*2,l,l)

        #embedding
        nodes_mask=position_idx.eq(0)
        token_mask=position_idx.ge(2)        
        inputs_embeddings=self.encoder.roberta.embeddings.word_embeddings(inputs_ids)
        nodes_to_token_mask=nodes_mask[:,:,None]&token_mask[:,None,:]&attn_mask
        nodes_to_token_mask=nodes_to_token_mask/(nodes_to_token_mask.sum(-1)+1e-10)[:,:,None]
        avg_embeddings=torch.einsum("abc,acd->abd",nodes_to_token_mask,inputs_embeddings)
        inputs_embeddings=inputs_embeddings*(~nodes_mask)[:,:,None]+avg_embeddings*nodes_mask[:,:,None]    
        
        outputs = self.encoder.roberta(inputs_embeds=inputs_embeddings,attention_mask=attn_mask,position_ids=position_idx,token_type_ids=position_idx.eq(-1).long())[0]`

**Comment by zzxn** [comment]
I am also confused that in the CodeXGLUE repo, CodeBERT is finetuned by concatenating the source tokens instead of [CLS] rep, different from GraphCodeBERT of this repo. Any explanation? 🤔

**Comment by guoday** [comment]
sorry for late reply. Both CodeBERT and GraphCodeBERT concatenates [CLS] vector of two source code, and then feed the concatenated vector into a linear layer for binary classification. Please refer [here](https://github.com/microsoft/CodeBERT/blob/f01b02dfdae8e14f4a0ba61cbe92ba72060c7c76/GraphCodeBERT/clonedetection/model.py#L19-L25) and [here](https://github.com/microsoft/CodeXGLUE/blob/148597ebdc90c9fde66be0ccee89651f1c937d36/Code-Code/Clone-detection-BigCloneBench/code/model.py#L21-L27).

**Comment by tyhxyz** [comment]
这是来自QQ邮箱的假期自动回复邮件。您好，我最近正在休假中，无法亲自回复您的邮件。我将在假期结束后，尽快给您回复。

**Comment by zzxn** [comment]
> sorry for late reply. Both CodeBERT and GraphCodeBERT concatenates [CLS] vector of two source code, and then feed the concatenated vector into a linear layer for binary classification. Please refer [here](https://github.com/microsoft/CodeBERT/blob/f01b02dfdae8e14f4a0ba61cbe92ba72060c7c76/GraphCodeBERT/clonedetection/model.py#L19-L25) and [here](https://github.com/microsoft/CodeXGLUE/blob/148597ebdc90c9fde66be0ccee89651f1c937d36/Code-Code/Clone-detection-BigCloneBench/code/model.py#L21-L27).

OK, thanks!
